packages/dependency-extraction-webpack-plugin/tsconfig.json

```diff
@@ -1,8 +0,0 @@
-{
-	"extends": "../../tsconfig.base.json",
-	"compilerOptions": {
-		"rootDir": "lib",
-		"declarationDir": "build-types"
-	},
-	"include": [ "lib/**/*" ]
-}
```

Comment:
I'm OK with removing this in the name of progress, but I'm also willing to help if we can keep the generated types.

The API of the plugins will remain the same, correct? In that case maybe we can generate and commit the types (compile and move build-types to types).

We need to delete `types` from the package.json if we remove them, or updated it if we move and commit the types.

Reply:
I've been thinking about the why's of types in this plugin and my conclusion is:

Actually implementing the plugin in TypeScript, and ensuring its source is typed correctly, is super difficult. To make the plugin compatible with both webpack 4 and 5, I'd need to import two different typings into the same source, and use them at the same time:
```js
if ( webpack.version.startsWith( '4.' ) ) {
  // do stuff with webpack 4 types
} else {
  // do stuff with webpack 5 types
}
```
I don't think that's even possible.

Also, the types published by webpack are incomplete. For example, v5 doesn't publish type for `ConcatenatedModule` which we need to descend into the concatenated module's `.modules` list: it can contain `ExternalModule`s that got concatenated into one moduled and removed from the bundle.

Then we need to be to check if a `Module` is an instance of `ExternalModule` or `ConcatenatedModule`, before accessing their `.externalType` or `.modules` fields type-safely. But we can't do
```js
if ( module instanceof ConcatenatedModule )
```
because the `ConcatenatedModule` constructor is not exported by webpack. We need the physical reference to the class at runtime, not just the type declaration.

We'd need to land several webpack patches like this before we could continue.

At the same time, what we really want is probably not implementing the plugin in TypeScript, but just to publish the types declaration so that _webpack config authors_ can write their config in TypeScript.

For that, a pair of `index.js` and `index.d.ts` files, without any build step, looks ideal to me. What do you think?

Reply:
That all seems reasonable and I'd be happy with what you propose 👍 

Thanks for sharing the detailed thoughts.